ga4gh/backends.py

```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
         for variantSetId in os.listdir(self._dataDir):
             relativePath = os.path.join(self._dataDir, variantSetId)
-            # A variant set is all files under a directory,
-            # so we skip non-directory files.
-            if os.path.isfile(relativePath):
-                continue
-            print(variantSetId)
-            self._variantSetMap[variantSetId] = variantSetClass(
-                variantSetId, relativePath)
-        self._variantSetIds = sorted(self._variantSetMap.keys())
+            if os.path.isdir(relativePath):
+                self._variantSetIdMap[variantSetId] = variantSetClass(
+                    variantSetId, relativePath)
+        self._variantSetIds = sorted(self._variantSetIdMap.keys())
 
-    def searchVariants(self, request):
-        assert len(request.variantSetIds) > 0
-        variantSetIndex = 0
-        if request.pageToken is not None:
-            # parse the pageToken and change what will be passed in
-            variantSetIndex, pageToken = request.pageToken.split(':')
-            variantSetIndex = int(variantSetIndex)
-            if pageToken == '':
-                pageToken = None  # None is represented as '' in the pageToken
-            else:
-                pageToken = int(pageToken)
-            request.pageToken = pageToken
-        variantSet = self._variantSetMap[
-            request.variantSetIds[variantSetIndex]]
-        response = variantSet.searchVariants(request)
-        # Add the index of the variant set of the next
-        # page to the nextPageToken
-        if response.nextPageToken is None:
-            if variantSetIndex < len(request.variantSetIds) - 1:
-                # if not, there are no more results, so leave the token as None
-                response.nextPageToken = '{0}:'.format(variantSetIndex + 1)
-                # this token will give results from the beginning of the
-                # next variantSet
-        else:
-            response.nextPageToken = "{0}:{1}".format(variantSetIndex,
-                                                      response.nextPageToken)
-        return response
+    def runSearchRequest(
+            self, requestStr, requestClass, responseClass, pageListName,
+            objectGenerator):
+        """
+        Runs the specified request. The request is a string containing
+        a JSON representation of an instance of the specified requestClass
+        in which the page list variable has the specified pageListName.
+        We return a string representation of an instance of the specified
+        responseClass in JSON format. Objects are filled into the page list
+        using the specified object generator, which must return
+        (object, nextPageToken) pairs, and be able to resume iteration from
+        any point using the nextPageToken attribute of the request object.
+        """
+        request = requestClass.fromJSON(requestStr)
```

Comment:
I would call validate here, and therefore either add a validate JSON string method to protocol or else pass in the JSON dict here.